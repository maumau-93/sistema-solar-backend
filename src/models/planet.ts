import IPlanet from './interfaces/I-planet';
import mongoose from 'mongoose';

/**
 * This interface is created to satisfy the inheritance of all the properties
 * of mongoose.Document class
 */
interface PlanetModel extends IPlanet, mongoose.Document {}

const Planet = new mongoose.Schema({
  name: {
    type: String,
    required: true,
    unique: true,
    minlength: 2,
    maxlength: 30,
    index: true
  },
  distanceToSun: {
    type: Number,
    required: true
  },
  angularSpeed: {
    type: Number,
    required: true
  },
  planetDirection: {
    type: Number,
    required: true,
    enum: [1, -1]
  }
});

const planetSchema = mongoose.model<PlanetModel>('Planet', Planet);

export { planetSchema, PlanetModel };
