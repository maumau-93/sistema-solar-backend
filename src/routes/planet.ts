import { Router } from 'express';
import createPlanet from '../api/planet';
import planetSchemaValidator from '../validators/planet';

const route = Router();

const planetRoutes: (app: Router) => void = (app: Router) => {
  app.use('/planet', route);
  route.post('', planetSchemaValidator, createPlanet);
};

export default planetRoutes;
