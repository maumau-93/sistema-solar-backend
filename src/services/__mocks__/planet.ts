import IPlanet from '../../models/interfaces/I-planet';

const createNewPlanet = (planet: IPlanet): any => {
  if (planet.name.length >= 2) return planet;
  throw new Error(planet.name);
};

const getInitialPlanets = (planetNames: string[]): any => {
  if (planetNames.length === 3) return ['Success'];
  throw new Error('Error occured');
};

export { createNewPlanet, getInitialPlanets };
