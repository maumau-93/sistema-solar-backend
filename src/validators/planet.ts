import { Validator } from 'jsonschema';
import { Request, Response, NextFunction } from 'express';
import { HTTP_CODES, PLANET_VALIDATION } from '../utils/global';
import IPlanet from '../models/interfaces/I-planet';

const validator = new Validator();

const planetSchema = {
  properties: {
    name: {
      type: 'string',
      minLength: 2,
      maxLength: 30
    },
    distanceToSun: {
      type: 'number'
    },
    angularSpeed: {
      type: 'number'
    },
    planetDirection: {
      type: 'number'
    }
  },
  required: ['name', 'distanceToSun', 'angularSpeed', 'planetDirection'],
  additionalProperties: false
};

const planetSchemaValidator = (
  req: Request,
  res: Response,
  next: NextFunction
): void => {
  const newPlanet = req.body.planet as IPlanet;
  const validatorResponse = validator.validate(newPlanet, planetSchema);
  if (validatorResponse.errors.length === 0) {
    req.body.planet = newPlanet;
    next();
  } else {
    res.status(HTTP_CODES.BAD_REQUEST).send({
      status: HTTP_CODES.BAD_REQUEST,
      message: PLANET_VALIDATION.INVALID_PLANET_OBJECT
    });
  }
};

export default planetSchemaValidator;
