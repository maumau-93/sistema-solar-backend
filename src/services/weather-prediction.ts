import env from '../config/env';
import events from '../events/events';
import IPlanet from '../models/interfaces/I-planet';
import IWeatherPrediction from '../models/interfaces/I-weather-prediction';
import Logger from '../config/loaders/logger';
import redisInstance from '../config/loaders/redis';
import { promisify } from 'util';
import { weatherPredictionSchema } from '../models/weather-prediction';
import {
  GET_WEATHER_MESSAGES,
  GET_DAY_MESSAGES,
  GET_MAX_PREC_RANGE,
  GET_MAX_PREC_HIST,
  GET_PERIOD_WEATHER_COUNT,
  UPDATE_MAX_PREC,
  WEATHER_CREATION_MESSAGES,
  REDIS_CHECKING,
  WEATHER_CACHE_POSTFIX
} from '../utils/global';
import {
  calculateCoordinates,
  determineWeatherForDay
} from '../utils/weather-prediction';

const getAsync = promisify(redisInstance.get).bind(redisInstance);

/**
 * Returns the count of periods for the specified weather
 * @param weather
 * @returns {Promise<number>}
 */
const getPeriodCountForWeather = async (weather: string): Promise<number> => {
  Logger.info(GET_PERIOD_WEATHER_COUNT.START);
  return await weatherPredictionSchema.countDocuments({
    weather,
    startingPeriod: true
  });
};

/**
 * Returns the weather prediction for a given day from the Redis cache
 * whenever possible.
 * @param day
 */
const getWeatherPredictionFromCache = async (day: number) => {
  Logger.info(REDIS_CHECKING.START_CACHE_CHECK);
  try {
    const weatherPrediction = await getAsync(`${day}${WEATHER_CACHE_POSTFIX}`);
    if (!weatherPrediction) {
      Logger.info(REDIS_CHECKING.DAY_NOT_CACHED);
      return null;
    }
    return JSON.parse(weatherPrediction) as IWeatherPrediction;
  } catch (e) {
    Logger.info(REDIS_CHECKING.ERROR_GETTING_FROM_REDIS);
    Logger.info(e);
    return null;
  }
};

/**
 * Returns the weather prediction for the specified day
 * @param day
 * @returns {Promise<IWeatherPrediction>}
 */
const getWeatherPrediction = async (
  day: number
): Promise<IWeatherPrediction> => {
  try {
    const cachedPrediction = await getWeatherPredictionFromCache(day);
    if (cachedPrediction) return cachedPrediction;
    const weatherPrediction = await weatherPredictionSchema.findOne({ day });
    if (!weatherPrediction) {
      throw new Error(GET_WEATHER_MESSAGES.GET_ERROR_NOT_FOUND);
    }
    env.eventEmitter.emit(events.weatherQuery.onQuery, day);
    return weatherPrediction;
  } catch (e) {
    Logger.info(GET_WEATHER_MESSAGES.GET_ERROR_FAILURE);
    Logger.error(e);
    throw e;
  }
};

/**
 * Returns the most recently processed weather prediction
 * @returns {Promise<IWeatherPrediction>}
 */
const getMostRecentProcessedDay = async (): Promise<IWeatherPrediction> => {
  try {
    const mostRecentProcessedDay = await weatherPredictionSchema
      .find()
      .sort({ day: -1 })
      .limit(1);
    return mostRecentProcessedDay[0];
  } catch (e) {
    Logger.info(GET_DAY_MESSAGES.GET_ERROR_FAILURE);
    Logger.error(e);
    throw e;
  }
};

/**
 * Returns the weather precition for the maximum precipitation in the given
 * range of days
 * @param startDay
 * @param endDay
 * @returns {Promise<IWeatherPrediction>}
 */
const getMaximumPrecipitationPredictionInRangeOfDays = async (
  startDay: number,
  endDay: number
): Promise<IWeatherPrediction> => {
  try {
    const maxPrecipitationDay = await weatherPredictionSchema
      .find({
        day: { $gte: startDay, $lte: endDay },
        precipitationPolygonPerimeter: { $exists: true }
      })
      .sort({ precipitationPolygonPerimeter: -1 })
      .limit(1);
    return maxPrecipitationDay[0];
  } catch (e) {
    Logger.info(GET_MAX_PREC_RANGE.ERROR);
    Logger.error(e);
    throw e;
  }
};

/**
 * Retruns the weather prediction for the most rainy day in history
 * @returns {Promise<IWeatherPrediction>}
 */
const getMaximumPrecipitationPredictionInHistory = async (): Promise<
  IWeatherPrediction
> => {
  try {
    const maxPrecipitationDayInHistory = await weatherPredictionSchema.find({
      maxPrecipitation: true
    });
    return maxPrecipitationDayInHistory[0];
  } catch (e) {
    Logger.info(GET_MAX_PREC_HIST.ERROR);
    Logger.error(e);
    throw e;
  }
};

/**
 * Updates the maxPrecipitation flag for the given weather prediction to the given value
 * @param weatherPrediction
 * @param value
 */
const setMaxPrecipitationFlag = async (
  day: number,
  value: boolean
): Promise<void> => {
  try {
    await weatherPredictionSchema.findOneAndUpdate(
      { day },
      { $set: { maxPrecipitation: value } }
    );
  } catch (e) {
    Logger.info(UPDATE_MAX_PREC.ERROR);
    Logger.error(e);
    throw e;
  }
};

/**
 * Creates the weather prediction for the days in the range [startDay, endDay] for the given universe of planets
 * @param startDay
 * @param endDay
 * @param planets
 * @returns {Promise<void>}
 */
const createNewWeatherPrediction = async (
  startDay: number,
  endDay: number,
  planets: IPlanet[]
): Promise<void> => {
  const promiseList = [];
  try {
    Logger.info(WEATHER_CREATION_MESSAGES.STARTING_CREATION);
    const mostRecentWeatherPrediction = await getMostRecentProcessedDay();
    let mostRecentWeather = mostRecentWeatherPrediction
      ? mostRecentWeatherPrediction.weather
      : '';
    for (let i = startDay; i <= endDay; i++) {
      const cartesianCoordinatesMap = calculateCoordinates(planets, i);
      const weatherForDay = determineWeatherForDay(
        Array.from(cartesianCoordinatesMap.values())
      );
      const weatherPrediction: IWeatherPrediction = {
        day: i,
        weather: weatherForDay.predictedWeather,
        precipitationPolygonPerimeter:
          weatherForDay.precipitationPolygonPerimeter,
        startingPeriod: weatherForDay.predictedWeather !== mostRecentWeather
      };
      mostRecentWeather = weatherForDay.predictedWeather;
      promiseList.push(weatherPredictionSchema.create(weatherPrediction));
    }
    await Promise.all(promiseList);
    env.eventEmitter.emit(
      events.weatherPrediction.onCreation,
      startDay,
      endDay
    );
  } catch (e) {
    Logger.info(WEATHER_CREATION_MESSAGES.ERROR_CREATING);
    Logger.error(e);
    throw e;
  }
};

export {
  createNewWeatherPrediction,
  getMostRecentProcessedDay,
  setMaxPrecipitationFlag,
  getMaximumPrecipitationPredictionInHistory,
  getMaximumPrecipitationPredictionInRangeOfDays,
  getPeriodCountForWeather,
  getWeatherPrediction
};
