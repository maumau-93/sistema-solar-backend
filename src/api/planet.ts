import { Request, Response } from 'express';
import { createNewPlanet } from '../services/planet';
import { HTTP_CODES, PLANET_CREATION_MESSAGES } from '../utils/global';
import Logger from '../config/loaders/logger';

/**
 * Controller function that creates a new Planet for the galaxy.
 * POST /api/planet
 * Visit models/planet:Planet to check body requirements for the new planet to be
 * inserted
 * @param req
 * @param res
 */
const createPlanet: (req: Request, res: Response) => Promise<void> = async (
  req: Request,
  res: Response
) => {
  try {
    Logger.debug('Calling create planet endpoint with %o', req.body.planet);
    const planetCreated = await createNewPlanet(req.body.planet);
    res.status(HTTP_CODES.OK).json(planetCreated);
  } catch (e) {
    res
      .status(HTTP_CODES.INTERNAL_SERVER_ERROR)
      .json(PLANET_CREATION_MESSAGES.ERROR_CREATING);
  }
};

export default createPlanet;
