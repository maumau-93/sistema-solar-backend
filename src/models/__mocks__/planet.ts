import IPlanet from '../interfaces/I-planet';
import { planetWithValidData } from '../../tests/data/planet';
import { PLANET_CREATION_MESSAGES } from '../../utils/global';

const create = (planetToCreate: IPlanet): IPlanet => {
  if (planetToCreate.name.length > 2) return planetToCreate;
  throw new Error(PLANET_CREATION_MESSAGES.ERROR_CREATING);
};

const find = (): IPlanet[] => {
  return [planetWithValidData];
};

const planetSchema = {
  create,
  find
};
const planetModel = {};

export { planetSchema, planetModel };
