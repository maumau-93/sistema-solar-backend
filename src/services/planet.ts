import IPlanet from '../models/interfaces/I-planet';
import Logger from '../config/loaders/logger';
import { planetSchema } from '../models/planet';

import {
  PLANET_CREATION_MESSAGES,
  GET_PLANETS_MESSAGES
} from '../utils/global';

/**
 * Creates a new planet given a planet object and returns it
 * @param planetToCreate
 * @returns {Promise<IPlanet>}
 */
const createNewPlanet = async (planetToCreate: IPlanet): Promise<IPlanet> => {
  try {
    Logger.info(PLANET_CREATION_MESSAGES.STARTING_CREATION);
    const planetCreated = await planetSchema.create(planetToCreate);
    Logger.info(PLANET_CREATION_MESSAGES.CREATION_FINISHED);
    return planetCreated;
  } catch (e) {
    Logger.info(PLANET_CREATION_MESSAGES.ERROR_CREATING);
    Logger.error(e);
    throw e;
  }
};

/**
 * Returns a list of planets given a list of names
 * @param planetNames
 * @returns {Promise<IPlanet[]>}
 */
const getInitialPlanets = async (planetNames: string[]): Promise<IPlanet[]> => {
  try {
    Logger.info(GET_PLANETS_MESSAGES.GET_STARTED);
    const initialPlanets = await planetSchema.find({
      name: { $in: planetNames }
    });
    Logger.info(GET_PLANETS_MESSAGES.GET_FINISHED);
    return initialPlanets;
  } catch (e) {
    Logger.info(GET_PLANETS_MESSAGES.GET_ERROR_FAILURE);
    Logger.error(e);
    throw e;
  }
};

export { createNewPlanet, getInitialPlanets };
