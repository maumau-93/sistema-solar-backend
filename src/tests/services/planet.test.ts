import { createNewPlanet, getInitialPlanets } from '../../services/planet';
import { planetWithValidData, planetWithInvalidName } from '../data/planet';

jest.mock('../../models/planet');

describe('createPlanet', () => {
  it('Should save successfully a new Planet', async () => {
    const result = await createNewPlanet(planetWithValidData);
    expect(result.name).toBe(planetWithValidData.name);
  });
  it('Should fail to create a new Planet', async () => {
    try {
      await createNewPlanet(planetWithInvalidName);
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });
});

describe('getInitialPlanets', () => {
  it('should retreive successfully the requested planet', async () => {
    const result = await getInitialPlanets(['myPlanet']);
    expect(result[0].name).toBe(planetWithValidData.name);
  });
});
