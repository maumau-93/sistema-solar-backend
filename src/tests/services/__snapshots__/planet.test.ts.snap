// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPlanet Should fail to create a new Planet 1`] = `[Error: Error while creating a new planet]`;
